### PR TITLE
Improve OIDC compliance: add JWKS support

### PR DESCRIPTION
**Pull request by Widcket**
### Changes

This update improves the SDK support for **OpenID Connect (OIDC)**. In particular, it adds support for **JSON Web Key Set (JWKS)**. Additional PRs will further expand the SDK support for OIDC.

#### What’s being added in this PR
- JWKS endpoint. This PR does not include tests for this (will be in an upcoming PR), as they’re in the Authentication spec and that file includes changes that belong in other PRs, and I’m trying to keep the PR split simple.
- JWK and JWKS models.
- Logic for creating a SecKey from a modulus and exponent, and tests for that.
- Supported algorithms and some of its tests. The remaining tests on this will be in the next PR (signature validation).
- JWT and JWK generators for testing. This includes logic from Guardian for exporting the data from a SecKey and parsing it in order to extract the modulus and exponent. I added comments in the code explaining what it does as it’s hard to follow, since it involves pointers and low level stuff.
- [JWTDecode](https://github.com/auth0/JWTDecode.swift) as a dependency.

All this logic is isolated ATM, as the integration will come in the last PR.

##### Additions
- A public method in the Authentication API to retrieve the tenant’s JWKS data.
- The JWKS and JWK models.
- An extension to decode Base64URL data. This was made public as it’s the counterpart of the already-public a0_encodeBase64URLSafe().
- A0SHA and A0RSA, ObjC classes from Guardian that implement a wrapper around some CommonCrypto functionality.
##### Changes
None.

### Testing

- [X] This change adds unit test coverage
- [X] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [X] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [X] All existing and new tests complete without errors

### CODE REVIEW

File: Auth0/JWKS.swift [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        guard let modulus = rsaModulus?.a0_decodeBase64URLSafe(),
+            let exponent = rsaExponent?.a0_decodeBase64URLSafe() else { return nil }
+        let encodedKey = encodeRSAPublicKey(modulus: [UInt8](modulus), exponent: [UInt8](exponent))
+        if #available(iOS 10, OSX 10.12, tvOS 10, watchOS 3, *) {
+            return generateRSAPublicKey(from: encodedKey)
+        }
+        let tag = "com.auth0.tmp.RSAPublicKey"
+        let keychain = A0SimpleKeychain()
+        guard keychain.setRSAPublicKey(data: encodedKey, forKey: tag) else { return nil }
+        return keychain.keyRefOfRSAKey(withTag: tag).takeRetainedValue()
```
Comment (Widcket) [inline Auth0/JWKS.swift]: This fails on macOS 10.11 (El Capitan), which happens to be the minimum supported version. So in that specific macOS version, the signature validation is skipped. Check https://github.com/auth0/Auth0.swift/pull/333/files#diff-2345af69c12dd8c93d94fdb71f422ba3R38.

In https://github.com/auth0/Auth0.swift/pull/333/files#diff-622b00adee0ac311c480485596339902R38 and https://github.com/auth0/Auth0.swift/pull/333/files#diff-b2bf319ef5a2f202615bba8e8cd28887R50, version checks were added to acomodate this.

File: Auth0/A0RSA.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// A0RSA.h
```
Comment (Widcket) [inline Auth0/A0RSA.h]: Wrapper around CommonCrypto, so it can be used from any Swift version.

File: Auth0/A0RSA.m [hunk 3]
```diff
@@ -0,0 +1,71 @@
+// A0RSA.m
```
Comment (Widcket) [inline Auth0/A0RSA.m]: Wrapper around CommonCrypto, so it can be used from any Swift version.

File: Auth0/A0SHA.m [hunk 4]
```diff
@@ -0,0 +1,55 @@
+// A0SHA.m
```
Comment (Widcket) [inline Auth0/A0SHA.m]: Wrapper around CommonCrypto, so it can be used from any Swift version.

File: Auth0/A0SimpleKeychain+RSAPublicKey.swift [hunk 5]
```diff
@@ -0,0 +1,40 @@
+// A0SimpleKeychain+RSAPublicKey.swift
```
Comment (Widcket) [inline Auth0/A0SimpleKeychain+RSAPublicKey.swift]: Extension to save a RSA key in the Keychain.

File: Auth0/Array+Encode.swift [hunk 6]
```diff
@@ -0,0 +1,40 @@
+// Array+Encode.swift
```
Comment (Widcket) [inline Auth0/Array+Encode.swift]: Extension to der-encode a byte array.

File: Auth0/Auth0Authentication.swift [hunk 7]
```diff
@@ -59,7 +59,13 @@ struct Auth0Authentication: Authentication {
             ]
```
Comment (Widcket) [inline Auth0/Auth0Authentication.swift]: Authentication API implementation.

File: Auth0/Authentication.swift [hunk 8]
```diff
@@ -434,7 +434,7 @@ public protocol Authentication: Trackable, Loggable {
     func startPasswordless(phoneNumber: String, type: PasswordlessType, connection: String) -> Request<Void, AuthenticationError>
```
Comment (Widcket) [inline Auth0/Authentication.swift]: Authentication API protocol.

File: Auth0/Handlers.swift [hunk 9]
```diff
@@ -35,6 +35,22 @@ func plainJson(from response: Response<AuthenticationError>, callback: Request<[
     }
```
Comment (Widcket) [inline Auth0/Handlers.swift]: JWKS request handler.

File: Auth0/JWKS.swift [hunk 10]
```diff
@@ -0,0 +1,91 @@
+// JWKS.swift
```
Comment (Widcket) [inline Auth0/JWKS.swift]: JWKS and JWK models.

File: Auth0/JWKS.swift [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        case certChain = "x5c"
+        case rsaModulus = "n"
+        case rsaExponent = "e"
+    }
+}
+
+extension JWK {
+    var rsaPublicKey: SecKey? {
+        guard let modulus = rsaModulus?.a0_decodeBase64URLSafe(),
+            let exponent = rsaExponent?.a0_decodeBase64URLSafe() else { return nil }
```
Comment (lbalmaceda) [inline Auth0/JWKS.swift]: would this "else return nil" apply to the modulus or only to the exponent as well?

File: Auth0/JWKS.swift [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        let encodedKey = encodeRSAPublicKey(modulus: [UInt8](modulus), exponent: [UInt8](exponent))
+        if #available(iOS 10, OSX 10.12, tvOS 10, watchOS 3, *) {
+            return generateRSAPublicKey(from: encodedKey)
+        }
+        let tag = "com.auth0.tmp.RSAPublicKey"
```
Comment (lbalmaceda) [inline Auth0/JWKS.swift]: should this be a constant?

File: Auth0/JWTAlgorithm.swift [hunk 13]
```diff
@@ -0,0 +1,48 @@
+// JWTAlgorithm.swift
```
Comment (Widcket) [inline Auth0/JWTAlgorithm.swift]: Supported algorithms.

File: Auth0/String+URLSafe.swift [hunk 14]
```diff
@@ -0,0 +1,35 @@
+// String+URLSafe.swift
```
Comment (Widcket) [inline Auth0/String+URLSafe.swift]: Extension to decode data encoded in Base64Url.

Comment (Widcket) [reply Auth0/String+URLSafe.swift]: It's public because it's the counterpart of the already-public `a0_encodeBase64URLSafe()`.

File: Auth0/JWTAlgorithm.swift [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+enum JWTAlgorithm: String {
+    case rs256 = "RS256"
+    case hs256 = "HS256"
+    
+    func verify(_ jwt: JWT, using jwk: JWK) -> Bool {
+        let separator = "."
+        let parts = jwt.string.components(separatedBy: separator).dropLast().joined(separator: separator)
+        guard let data = parts.data(using: .utf8),
+            let signature = jwt.signature?.a0_decodeBase64URLSafe(),
+            !signature.isEmpty else { return false }
```
Comment (lbalmaceda) [inline Auth0/JWTAlgorithm.swift]: does this mean that if the signature is empty the verification fails gracefully?

File: Auth0/JWTAlgorithm.swift [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        switch self {
+        case .rs256:
+            #if os(OSX)
+            guard #available(OSX 10.12, *) else { return true }
+            #endif
+            guard let publicKey = jwk.rsaPublicKey,
+                let sha256 = A0SHA(algorithm: "sha256"),
```
Comment (lbalmaceda) [inline Auth0/JWTAlgorithm.swift]: do we need to pass the hashing algorithm, knowing we will only support one?

File: Auth0Tests/Extensions.swift [hunk 17]
```diff
@@ -0,0 +1,45 @@
+// Extensions.swift
```
Comment (Widcket) [inline Auth0Tests/Extensions.swift]: Extensions part of the test targets that support the JWT and JWKS generators.

File: Auth0Tests/Generators.swift [hunk 18]
```diff
@@ -0,0 +1,198 @@
+// Generators.swift
```
Comment (Widcket) [inline Auth0Tests/Generators.swift]: JWT and JWKS generators.

File: Auth0/String+URLSafe.swift [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import Foundation
+
+public extension String {
+    func a0_decodeBase64URLSafe() -> Data? {
+        let paddingLength = 4 - count % 4
```
Comment (lbalmaceda) [inline Auth0/String+URLSafe.swift]: what is this magic number? maybe move to a constant, give it a proper name/description?

File: Auth0Tests/Extensions.swift [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    func export() -> Data {
+        return SecKeyCopyExternalRepresentation(self, nil)! as Data
+    }
+}
+
+extension JWTAlgorithm {
+    func sign(value: Data, key: SecKey = TestKeys.rsaPrivate) -> Data {
+        switch self {
+        case .rs256:
+            let sha256 = A0SHA(algorithm: "sha256")!
```
Comment (lbalmaceda) [inline Auth0Tests/Extensions.swift]: consider moving `"sha256"` inside the `A0SHA` class

File: Auth0Tests/Extensions.swift [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+            let rsa = A0RSA(key: key)!
+            
+            return rsa.sign(sha256.hash(value))
+        case .hs256: return value
```
Comment (lbalmaceda) [inline Auth0Tests/Extensions.swift]: If I got this right, signing with HS256 is not signing but returning the data untouched?

File: Auth0Tests/Generators.swift [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+    
+    return encodeJWTPart(from: bodyDict)
+}
+
+func generateJWT(alg: String = JWTAlgorithm.rs256.rawValue,
+                 kid: String = "key123",
+                 iss: String? = "https://tokens-test.auth0.com/",
+                 sub: String? = "auth0|123456789",
+                 aud: [String]? = ["tokens-test-123"],
+                 exp: Date? = Date().addingTimeInterval(100000),
```
Comment (lbalmaceda) [inline Auth0Tests/Generators.swift]: magic numbers

File: Auth0Tests/Generators.swift [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+    else if let algorithm = JWTAlgorithm(rawValue: alg) {
+        signaturePart = algorithm.sign(value: signableParts.data(using: .utf8)!).a0_encodeBase64URLSafe()!
+    }
+    
+    return try! decode(jwt: "\(signableParts).\(signaturePart)")
+}
+
+// MARK: - JWK
+
+private func calculateLength(from bytes: UnsafePointer<UInt8>) -> (UnsafePointer<UInt8>, Int) {
```
Comment (lbalmaceda) [inline Auth0Tests/Generators.swift]: you might want to at least comment all these helper functions. With the byte manipulation in place, they are difficult to follow.

File: Auth0Tests/Generators.swift [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+    guard bytes.pointee == 0x02 else { return nil }
+    
+    let (valueBytes, valueLength) = calculateLength(from: bytes + 1)
+    let data = Data(bytes: valueBytes, count: valueLength)
+    let pointer = valueBytes + valueLength
+    
+    return (pointer, data)
+}
+
+func generateRSAJWK(from publicKey: SecKey = TestKeys.rsaPublic, keyId: String = "key123") -> JWK {
```
Comment (lbalmaceda) [inline Auth0Tests/Generators.swift]: the second time I see this `key123` value. Seems a good candidate for a constant value

File: Auth0Tests/JWKSpec.swift [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+                    it("should generate a RSA public key") {
+                        let publicKey = jwk.rsaPublicKey!
+                        let keyAttributes = SecKeyCopyAttributes(publicKey) as! [String: Any]
+                        
+                        expect(keyAttributes[String(kSecAttrKeyType)] as? String).to(equal(String(kSecAttrKeyTypeRSA)))
+                    }
+                }
+            }
+            
+            context("unsuccessful generation") {
```
Comment (lbalmaceda) [inline Auth0Tests/JWKSpec.swift]: are exponent and modulus the only reasons why it could fail? What about having a different usage or algorithm? I don't expect all the cases to be covered, but those 2 I'd say are a must

File: Auth0Tests/JWTAlgorithmSpec.swift [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+            context("RS256") {
+                let alg = "RS256"
+                
+                it("should return true with a correct RS256 signature") {
+                    let jwt = generateJWT(alg: alg)
+                    
+                    expect(JWTAlgorithm.rs256.verify(jwt, using: jwk)).to(beTrue())
+                }
+                
+                it("should return false with an empty signature") {
```
Comment (lbalmaceda) [inline Auth0Tests/JWTAlgorithmSpec.swift]: this answers one of my questions above 👍

File: Auth0Tests/JWTAlgorithmSpec.swift [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+                    
+                    expect(JWTAlgorithm.hs256.verify(jwt, using: jwk)).to(beTrue())
+                }
+                
+                it("should return false with an empty signature") {
+                    let jwt = generateJWT(alg: alg, signature: "")
+                    
+                    expect(JWTAlgorithm.hs256.verify(jwt, using: jwk)).to(beFalse())
+                }
+            }
```
Comment (lbalmaceda) [inline Auth0Tests/JWTAlgorithmSpec.swift]: what file is checking that any other algorithm name returns false as well?

File: Auth0/JWKS.swift (hunk 11 above)
Comment (Widcket) [reply Auth0/JWKS.swift]: Would apply to both of them. It's equivalent to the following if:
```swift
    if rsaModulus?.a0_decodeBase64URLSafe() != nil && rsaExponent?.a0_decodeBase64URLSafe() != nil {}
```

Comment (Widcket) [reply Auth0/JWKS.swift]: Would apply to both of them. It's equivalent to the following if:
```swift
    if rsaModulus?.a0_decodeBase64URLSafe() != nil && rsaExponent?.a0_decodeBase64URLSafe() != nil {
        let modulus = rsaModulus!.a0_decodeBase64URLSafe()!
        let exponent = rsaExponent!.a0_decodeBase64URLSafe()!
}

```

File: Auth0Tests/JWTAlgorithmSpec.swift (hunk 27 above)
Comment (Widcket) [reply Auth0Tests/JWTAlgorithmSpec.swift]: It's going to be in the next PR.

File: Auth0/JWTAlgorithm.swift (hunk 15 above)
Comment (Widcket) [reply Auth0/JWTAlgorithm.swift]: Yes

File: Auth0/JWTAlgorithm.swift (hunk 16 above)
Comment (Widcket) [reply Auth0/JWTAlgorithm.swift]: That's how the implementation from Guardian is written, and I didn't want to touch those crypto ObjC files.

File: Auth0/JWKS.swift (hunk 12 above)
Comment (Widcket) [reply Auth0/JWKS.swift]: I don't think so, since this is only used in this particular computed property of this particular extension.

File: Auth0Tests/Generators.swift (hunk 23 above)
Comment (Widcket) [reply Auth0Tests/Generators.swift]: I'll add some comments.

File: Auth0Tests/Extensions.swift (hunk 21 above)
Comment (Widcket) [reply Auth0Tests/Extensions.swift]: Yes, exactly.

File: Auth0Tests/JWKSpec.swift (hunk 25 above)
Comment (Widcket) [reply Auth0Tests/JWKSpec.swift]: The usage is not used at all in the `.rsaPublicKey` implementation. All that's needed is valid values in the properties `rsaModulus` and `rsaExponent`.

File: Auth0Tests/Generators.swift (hunk 22 above)
Comment (Widcket) [reply Auth0Tests/Generators.swift]: All the default param values in that method are sane defaults for the tests. What that particular line does is create a date 100000 milliseconds in the future.

File: Auth0Tests/Generators.swift (hunk 23 above)
Comment (Widcket) [inline Auth0Tests/Generators.swift]: This method calculates the length of a given DER triplet. It starts at the triplet's base address, the proceeds from there.

Comment (Widcket) [reply Auth0Tests/Generators.swift]: It returns a pointer to the end of the triplet plus its length.

File: Auth0Tests/Generators.swift [hunk 28]
```diff
@@ -0,0 +1,198 @@
...
+    guard bytes.pointee > 0x7f else { return (bytes + 1, Int(bytes.pointee)) }
+
+    let count = Int(bytes.pointee & 0x7f)
+    let length = (1...count).reduce(0) { ($0 << 8) + Int(bytes[$1]) }
+    let pointer = bytes + (1 + count)
+    
+    return (pointer, length)
+}
+
+private func extractData(from bytes: UnsafePointer<UInt8>) -> (UnsafePointer<UInt8>, Data)? {
```
Comment (Widcket) [inline Auth0Tests/Generators.swift]: This method extracts the data portion from a given DER triplet. It starts at the triplet's base address, then proceeds from there. It returns a pointer to the end of the triplet plus its data.

File: Auth0Tests/Generators.swift (hunk 23 above)
Comment (Widcket) [reply Auth0Tests/Generators.swift]: Done.

File: Auth0Tests/Generators.swift [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+    guard bytes.pointee == 0x02 else { return nil }
```
Comment (Widcket) [inline Auth0Tests/Generators.swift]: Checks that this is an integer triplet.

File: Auth0Tests/JWKSpec.swift (hunk 25 above)
Comment (Widcket) [reply Auth0Tests/JWKSpec.swift]: I can add a check in the guard clause of `.rsaPublicKey ` to return nil if the algorithm != "RSA" and usage != "sig".

Comment (Widcket) [reply Auth0Tests/JWKSpec.swift]: After all, it's an internal (package-private) property.

File: Auth0/String+URLSafe.swift (hunk 19 above)
Comment (Widcket) [reply Auth0/String+URLSafe.swift]: Base64 requires the string length to be a multiple of four. I can add a let constant in the line above with it.

File: Auth0Tests/Generators.swift (hunk 24 above)
Comment (Widcket) [reply Auth0Tests/Generators.swift]: I'll add a constant.

File: Auth0Tests/Extensions.swift (hunk 20 above)
Comment (Widcket) [reply Auth0Tests/Extensions.swift]: I'll add another initializer to `A0SHA` with `sha256 ` as the default alg.

File: Auth0/String+URLSafe.swift (hunk 19 above)
Comment (Widcket) [reply Auth0/String+URLSafe.swift]: Added in 123af59498094926539a355378c5ce78efa7a1bb.

File: Auth0Tests/Extensions.swift (hunk 20 above)
Comment (Widcket) [reply Auth0Tests/Extensions.swift]: Done in 123af59498094926539a355378c5ce78efa7a1bb.

File: Auth0Tests/Generators.swift (hunk 23 above)
Comment (Widcket) [reply Auth0Tests/Generators.swift]: Comment added in 123af59498094926539a355378c5ce78efa7a1bb.

File: Auth0Tests/Generators.swift (hunk 28 above)
Comment (Widcket) [reply Auth0Tests/Generators.swift]: Comment added in 123af59498094926539a355378c5ce78efa7a1bb.

File: Auth0Tests/Generators.swift (hunk 29 above)
Comment (Widcket) [reply Auth0Tests/Generators.swift]: Comment added in 123af59498094926539a355378c5ce78efa7a1bb.

File: Auth0Tests/Generators.swift (hunk 24 above)
Comment (Widcket) [reply Auth0Tests/Generators.swift]: Added in 123af59498094926539a355378c5ce78efa7a1bb.

File: Auth0Tests/JWKSpec.swift (hunk 25 above)
Comment (Widcket) [reply Auth0Tests/JWKSpec.swift]: Done in 123af59498094926539a355378c5ce78efa7a1bb. Thanks for spotting it!

**Comment by cocojoe** [comment]
@Widcket The JWTDecode framework files shouldn't be part of the repo, just as SimpleKeyChain framework files are not.

**Comment by Widcket** [comment]
@cocojoe removed.

File: Auth0Tests/Generators.swift (hunk 22 above)
Comment (lbalmaceda) [reply Auth0Tests/Generators.swift]: I get that, but let's give it a name or at least a comment on the side telling what is the intention. e.g. `// tomorrow` or `addingTimeInterval(ONE_DAY_IN_MS)`. Helps improving the readability

File: Auth0Tests/JWKSpec.swift (hunk 25 above)
Comment (lbalmaceda) [reply Auth0Tests/JWKSpec.swift]: @Widcket for clarification. The `use` key is optional according to the spec, but when present it should be used to check whether the key you are reading can be used for the operation you want to perform. 

For this particular implementation, it means that if `use!='sig'` then the key is invalid and should not be used.

Comment (Widcket) [reply Auth0Tests/JWKSpec.swift]: Then I should update the implementation, because I added a check in the guard clause testing if usage == "sig". I'll need to split it into a separate check, because `use` is optional, and only if present should be part of the validation. Thanks for the clarification.

File: Auth0Tests/JWTAlgorithmSpec.swift (hunk 27 above)
Comment (lbalmaceda) [reply Auth0Tests/JWTAlgorithmSpec.swift]: Should that be noted on a `//TODO` comment for posterity?

File: Auth0/JWKS.swift [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+        case certThumbprint = "x5t"
+        case certChain = "x5c"
+        case rsaModulus = "n"
+        case rsaExponent = "e"
+    }
+}
+
+extension JWK {
+    var rsaPublicKey: SecKey? {
+        if let usage = usage, usage != "sig" { return nil }
```
Comment (Widcket) [inline Auth0/JWKS.swift]: @lbalmaceda updated to reflect the fact that `use` is optional.

File: Auth0Tests/JWTAlgorithmSpec.swift (hunk 27 above)
Comment (Widcket) [reply Auth0Tests/JWTAlgorithmSpec.swift]: No need, since it's already done.

File: Auth0/Authentication.swift [hunk 31]
```diff
@@ -434,7 +434,7 @@ public protocol Authentication: Trackable, Loggable {
...
 
     /**
-     Returns token information by performing a request to /tokeninfo endpoint
```
Comment (cocojoe) [inline Auth0/Authentication.swift]: again would avoid PR noise with formatting in the future

**Comment by Widcket** [comment]
> Can you update the description as discussed internally as lot going on. Also new lib added.

Done.
